File: 백준/Silver/11052. 카드 구매하기/카드 구매하기.py
n = int(input())

arr = list(map(int, input().split()))


dp = [0] * (n + 1)

dp[1] = arr[0]

for x in range(1, n + 1):
    for y in range(1, x + 1):
        dp[x] = max(dp[x], dp[x - y] + arr[y - 1])
print(dp[n])

File: 백준/Silver/11052. 카드 구매하기/카드 구매하기.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        int n = Integer.parseInt(br.readLine());
        
        StringTokenizer st = new StringTokenizer(br.readLine());
        int[] arr = new int[n];
        for (int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }
        
        int[] dp = new int[n + 1];
        dp[1] = arr[0];
        
        for (int x = 1; x <= n; x++) {
            for (int y = 1; y <= x; y++) {
                dp[x] = Math.max(dp[x], dp[x - y] + arr[y - 1]);
            }
        }
        
        System.out.println(dp[n]);
    }
}